text/2021-04-23-builtin-gc-worker.md

```diff
@@ -0,0 +1,131 @@
+# Builtin GC worker
+
+## Summary
+
+Move TiKV MVCC GC worker from TiDB into PD and make it general for use without TiDB.
+
+## Motivation
+
+GC worker is an important component for TiKV that deletes outdated MVCC data so as to not explode the storage. But currently, the GC worker is implemented in TiDB, which makes TiKV not usable without TiDB.
+
+## Background
+
+According to the [documentation](https://docs.pingcap.com/tidb/stable/garbage-collection-overview) for the **current** GC worker, the GC worker processes the following steps in a GC cycle:
+
+1. Calculate the minimal timestamp among all living SQL sessions, earlier than which are safe to GC, called `Service Safepoint`.
+2. [Upload](https://github.com/pingcap/kvproto/blob/8ecb5e46d7f5f7952a1a8d262b54f61dc8de1ef3/proto/pdpb.proto#L73) the `Service Safepoint` to PD, and get the minimal `Service Safepoint` among all Tools (e.g. CDC) from the response.
+3. Resolve all locks whose timestamp is earlier than the `Minimal Service Safepoint`.
+4. [Push](https://github.com/pingcap/kvproto/blob/8ecb5e46d7f5f7952a1a8d262b54f61dc8de1ef3/proto/pdpb.proto#L71) the `Minimal Service Safepoint` as `GCSafepoint` to PD.
+
+After the `GCSafepoint` being uploaded, TiKV will pull it from PD by interval, then TiKV will automatically delete all records earlier than `GCSafepoint`.
+
+Yet another tricky part is that TiDB has a shortcut to GC a range of keys after `DROP TABLE` or `DROP INDEX`: parallel with step 3 and step 4, TiDB will directly invoke `UnsafeDeleteRange` to TiKV which will delete the range of key-values regardless of locks or timestamp. To make it safe, TiDB will guarantee that the deleted range will never be used again.
+
+## Detailed design
+
+Three new concepts will be introduced by this proposal:
+
+1. `GC Barrier`
+
+    PD guarantees that the GC process will not affect the data with a timestamp later than any `GC Barrier`. TiKV users (e.g. TiDB / CDC / Client) can put GC Barriers into PD so as to make sure the data they are using will not be GC.
```

Comment:
> PD guarantees that the GC process will not affect the data with a timestamp later than any `GC Barrier`.

Strictly speaking, *all snapshots with a TS after GC Barrier should be consistent*. Which means, not only the versions with a timestamp later than GC Barrier should be kept, but also the latest version before GC Barrier.

This description is also used when we are talking about safepoints.

Reply:
I know it, maybe `snapshot ts` could make it clear.

Reply:
updated